// vk-io/lib/auth/implicit-flow-groups.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _debug = _interopRequireDefault(require("debug"));

var _cheerio = require("cheerio");

var _url = require("url");

var _implicitFlow = _interopRequireDefault(require("./implicit-flow"));

var _errors = require("../errors");

var _constants = require("../util/constants");

var _helpers = require("./helpers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const debug = (0, _debug.default)('vk-io:auth:implicit-flow-user');
const {
  AUTHORIZATION_FAILED
} = _errors.authErrors;

class ImplicitFlowGroups extends _implicitFlow.default {
  /**
   * Constructor
   *
   * @param {VK}     vk
   * @param {Object} options
   */
  constructor(vk, options) {
    super(vk, options);
    let {
      groups = null
    } = options;

    if (groups === null) {
      throw Error('Groups list must have');
    }

    if (!Array.isArray(groups)) {
      groups = [groups];
    }

    this.groups = groups.map(group => {
      if (typeof group !== 'number') {
        group = Number(group);
      }

      if (group < 0) {
        group = -group;
      }

      return group;
    });
  }
  /**
   * Returns permission page
   *
   * @param {Array} groups
   *
   * @return {Response}
   */


  getPermissionsPage() {
    const {
      app
    } = this;
    let {
      scope
    } = this;

    if (scope === 'all' || scope === null) {
      scope = (0, _helpers.getAllGroupsPermissions)();
    } else if (typeof scope !== 'number') {
      scope = (0, _helpers.getGroupsPermissionsByName)(scope);
    }

    debug('auth scope %s', scope);
    const params = new _url.URLSearchParams({
      group_ids: this.groups.join(','),
      redirect_uri: _constants.CALLBACK_BLANK,
      response_type: 'token',
      display: 'page',
      v: _constants.API_VERSION,
      client_id: app,
      scope
    });
    const url = new _url.URL(`https://oauth.vk.com/authorize?${params}`);
    return this.fetch(url, {
      method: 'GET'
    });
  }
  /**
   * Starts authorization
   *
   * @return {Promise<Array>}
   */


  async run() {
    const {
      response
    } = await super.run();
    let {
      hash
    } = new _url.URL(response.url);

    if (hash.startsWith('#')) {
      hash = hash.substring(1);
    }

    const params = new _url.URLSearchParams(hash);

    if (params.has('error')) {
      throw new _errors.AuthError({
        message: `Failed passed grant access: ${params.get('error_description') || 'Unknown error'}`,
        code: AUTHORIZATION_FAILED
      });
    }

    let expires = params.get('expires_in');

    if (expires !== null) {
      expires = Number(expires);
    }

    const tokens = [];

    for (const [name, value] of params) {
      if (!name.startsWith('access_token_')) {
        continue;
      }
      /* Example group access_token_XXXXX */


      const [,, group] = name.split('_');
      tokens.push({
        group: Number(group),
        token: value,
        expires
      });
    }

    return tokens;
  }

}

exports.default = ImplicitFlowGroups;
