// vk-io/lib/auth/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _util = require("util");

var _direct = _interopRequireDefault(require("./direct"));

var _implicitFlowUser = _interopRequireDefault(require("./implicit-flow-user"));

var _implicitFlowGroups = _interopRequireDefault(require("./implicit-flow-groups"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Auth {
  /**
   * Constructor
   *
   * @param {VK} vk
   */
  constructor(vk) {
    this.vk = vk;
  }
  /**
   * Returns custom tag
   *
   * @return {string}
   */


  get [Symbol.toStringTag]() {
    return 'Auth';
  }
  /**
   * Standalone authorization with login & password
   *
   * @return {ImplicitFlowUser}
   */


  implicitFlowUser(options = {}) {
    return new _implicitFlowUser.default(this.vk, options);
  }
  /**
   * Standalone authorization with login & password for group
   *
   * @param {mixed}  groups
   * @param {Object} options
   *
   * @return {ImplicitFlowGroup}
   */


  implicitFlowGroups(groups, options = {}) {
    return new _implicitFlowGroups.default(this.vk, Object.assign({}, options, {
      groups
    }));
  }
  /**
   * Direct authorization with login & login in user application
   *
   * @return {DirectAuth}
   */


  direct() {
    const {
      app,
      key
    } = this.vk.options;
    return new _direct.default(this.vk, {
      app,
      key
    });
  }
  /**
   * Direct authorization with login & login in android application
   *
   * @return {DirectAuth}
   */


  androidApp() {
    return new _direct.default(this.vk, {
      app: 2274003,
      key: 'hHbZxrka2uZ6jB1inYsH'
    });
  }
  /**
   * Direct authorization with login & login in windows application
   *
   * @return {DirectAuth}
   */


  windowsApp() {
    return new _direct.default(this.vk, {
      app: 3697615,
      key: 'AlVXZFMUqyrnABp8ncuU'
    });
  }
  /**
   * Direct authorization with login & login in windows phone application
   *
   * @return {DirectAuth}
   */


  windowsPhoneApp() {
    return new _direct.default(this.vk, {
      app: 3502557,
      key: 'PEObAuQi6KloPM4T30DV'
    });
  }
  /**
   * Direct authorization with login & login in iphone application
   *
   * @return {DirectAuth}
   */


  iphoneApp() {
    return new _direct.default(this.vk, {
      app: 3140623,
      key: 'VeWdmVclDCtn6ihuP1nt'
    });
  }
  /**
   * Direct authorization with login & login in ipad application
   *
   * @return {DirectAuth}
   */


  ipadApp() {
    return new _direct.default(this.vk, {
      app: 3682744,
      key: 'mY6CDUswIVdJLCD3j15n'
    });
  }
  /**
   * Custom inspect object
   *
   * @param {?number} depth
   * @param {Object}  options
   *
   * @return {string}
   */


  [_util.inspect.custom](depth, options) {
    const {
      name
    } = this.constructor;
    return `${options.stylize(name, 'special')} {}`;
  }

}

exports.default = Auth;
