// vk-io/lib/errors/api.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _vk = _interopRequireDefault(require("./vk"));

var _constants = require("../util/constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const {
  CAPTCHA_REQUIRED,
  USER_VALIDATION_REQUIRED,
  CONFIRMATION_REQUIRED
} = _constants.apiErrors;

class APIError extends _vk.default {
  /**
   * Constructor
   *
   * @param {Object} payload
   */
  constructor(payload) {
    const code = Number(payload.error_code);
    const message = `Code №${code} - ${payload.error_msg}`;
    super({
      code,
      message
    });
    this.params = payload.request_params;

    if (code === CAPTCHA_REQUIRED) {
      this.captchaSid = Number(payload.captcha_sid);
      this.captchaImg = payload.captcha_img;
    } else if (code === USER_VALIDATION_REQUIRED) {
      this.redirectUri = payload.redirect_uri;
    } else if (code === CONFIRMATION_REQUIRED) {
      this.confirmationText = payload.confirmation_text;
    }
  }

}

exports.default = APIError;
