// vk-io/lib/errors/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "VKError", {
  enumerable: true,
  get: function () {
    return _vk.default;
  }
});
Object.defineProperty(exports, "APIError", {
  enumerable: true,
  get: function () {
    return _api.default;
  }
});
Object.defineProperty(exports, "AuthError", {
  enumerable: true,
  get: function () {
    return _auth.default;
  }
});
Object.defineProperty(exports, "UploadError", {
  enumerable: true,
  get: function () {
    return _upload.default;
  }
});
Object.defineProperty(exports, "CollectError", {
  enumerable: true,
  get: function () {
    return _collect.default;
  }
});
Object.defineProperty(exports, "UpdatesError", {
  enumerable: true,
  get: function () {
    return _updates.default;
  }
});
Object.defineProperty(exports, "ExecuteError", {
  enumerable: true,
  get: function () {
    return _execute.default;
  }
});
Object.defineProperty(exports, "SnippetsError", {
  enumerable: true,
  get: function () {
    return _snippets.default;
  }
});
Object.defineProperty(exports, "StreamingRuleError", {
  enumerable: true,
  get: function () {
    return _streamingRule.default;
  }
});
Object.defineProperty(exports, "apiErrors", {
  enumerable: true,
  get: function () {
    return _constants.apiErrors;
  }
});
Object.defineProperty(exports, "authErrors", {
  enumerable: true,
  get: function () {
    return _constants.authErrors;
  }
});
Object.defineProperty(exports, "uploadErrors", {
  enumerable: true,
  get: function () {
    return _constants.uploadErrors;
  }
});
Object.defineProperty(exports, "updatesErrors", {
  enumerable: true,
  get: function () {
    return _constants.updatesErrors;
  }
});
Object.defineProperty(exports, "collectErrors", {
  enumerable: true,
  get: function () {
    return _constants.collectErrors;
  }
});
Object.defineProperty(exports, "snippetsErrors", {
  enumerable: true,
  get: function () {
    return _constants.snippetsErrors;
  }
});
exports.default = void 0;

var _vk = _interopRequireDefault(require("./vk"));

var _api = _interopRequireDefault(require("./api"));

var _auth = _interopRequireDefault(require("./auth"));

var _upload = _interopRequireDefault(require("./upload"));

var _collect = _interopRequireDefault(require("./collect"));

var _updates = _interopRequireDefault(require("./updates"));

var _execute = _interopRequireDefault(require("./execute"));

var _snippets = _interopRequireDefault(require("./snippets"));

var _streamingRule = _interopRequireDefault(require("./streaming-rule"));

var _constants = require("../util/constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = _vk.default;
exports.default = _default;
