// vk-io/lib/structures/contexts/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Context", {
  enumerable: true,
  get: function () {
    return _context.default;
  }
});
Object.defineProperty(exports, "VoteContext", {
  enumerable: true,
  get: function () {
    return _vote.default;
  }
});
Object.defineProperty(exports, "TypingContext", {
  enumerable: true,
  get: function () {
    return _typing.default;
  }
});
Object.defineProperty(exports, "MessageContext", {
  enumerable: true,
  get: function () {
    return _message.default;
  }
});
Object.defineProperty(exports, "WallPostContext", {
  enumerable: true,
  get: function () {
    return _wallPost.default;
  }
});
Object.defineProperty(exports, "StreamingContext", {
  enumerable: true,
  get: function () {
    return _streaming.default;
  }
});
Object.defineProperty(exports, "GroupUserContext", {
  enumerable: true,
  get: function () {
    return _groupUser.default;
  }
});
Object.defineProperty(exports, "UserOnlineContext", {
  enumerable: true,
  get: function () {
    return _userOnline.default;
  }
});
Object.defineProperty(exports, "DialogFlagsContext", {
  enumerable: true,
  get: function () {
    return _dialogFlags.default;
  }
});
Object.defineProperty(exports, "GroupUpdateContext", {
  enumerable: true,
  get: function () {
    return _groupUpdate.default;
  }
});
Object.defineProperty(exports, "GroupMemberContext", {
  enumerable: true,
  get: function () {
    return _groupMember.default;
  }
});
Object.defineProperty(exports, "MessageAllowContext", {
  enumerable: true,
  get: function () {
    return _messageAllow.default;
  }
});
Object.defineProperty(exports, "ReadMessagesContext", {
  enumerable: true,
  get: function () {
    return _readMessages.default;
  }
});
Object.defineProperty(exports, "MessageFlagsContext", {
  enumerable: true,
  get: function () {
    return _messageFlags.default;
  }
});
Object.defineProperty(exports, "CommentActionContext", {
  enumerable: true,
  get: function () {
    return _commentAction.default;
  }
});
Object.defineProperty(exports, "NewAttachmentsContext", {
  enumerable: true,
  get: function () {
    return _newAttachments.default;
  }
});
Object.defineProperty(exports, "RemovedMessagesContext", {
  enumerable: true,
  get: function () {
    return _removedMessages.default;
  }
});
exports.default = void 0;

var _context = _interopRequireDefault(require("./context"));

var _vote = _interopRequireDefault(require("./vote"));

var _typing = _interopRequireDefault(require("./typing"));

var _message = _interopRequireDefault(require("./message"));

var _wallPost = _interopRequireDefault(require("./wall-post"));

var _streaming = _interopRequireDefault(require("./streaming"));

var _groupUser = _interopRequireDefault(require("./group-user"));

var _userOnline = _interopRequireDefault(require("./user-online"));

var _dialogFlags = _interopRequireDefault(require("./dialog-flags"));

var _groupUpdate = _interopRequireDefault(require("./group-update"));

var _groupMember = _interopRequireDefault(require("./group-member"));

var _messageAllow = _interopRequireDefault(require("./message-allow"));

var _readMessages = _interopRequireDefault(require("./read-messages"));

var _messageFlags = _interopRequireDefault(require("./message-flags"));

var _commentAction = _interopRequireDefault(require("./comment-action"));

var _newAttachments = _interopRequireDefault(require("./new-attachments"));

var _removedMessages = _interopRequireDefault(require("./removed-messages"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = _context.default;
exports.default = _default;
