// vk-io/lib/structures/contexts/removed-messages.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _context = _interopRequireDefault(require("./context"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class RemovedMessagesContext extends _context.default {
  /**
   * Constructor
   *
   * @param {VK}     vk
   * @param {Arrray} update
   */
  constructor(vk, [eventId, peer, id]) {
    super(vk);
    this.payload = {
      peer_id: peer,
      id
    };
    this.type = 'removed_messages';
    this.subTypes = [eventId === 13 ? 'delete_messages' : 'restore_messages'];
  }
  /**
   * Checks that messages have been deleted
   *
   * @return {boolean}
   */


  isRemoved() {
    return this.subTypes.includes('delete_messages');
  }
  /**
   * Checks that messages have been restored
   *
   * @return {boolean}
   */


  isRecovery() {
    return this.subTypes.includes('restore_messages');
  }
  /**
   * Returns the identifier of the message
   *
   * @return {string}
   */


  getId() {
    return this.payload.id;
  }
  /**
   * Returns the peer ID
   *
   * @return {number}
   */


  getPeerId() {
    return this.payload.peer_id;
  }

}

exports.default = RemovedMessagesContext;
