// vk-io/lib/structures/contexts/vote.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _context = _interopRequireDefault(require("./context"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class VoteContext extends _context.default {
  /**
   * Constructor
   *
   * @param {VK}     vk
   * @param {Object} payload
   */
  constructor(vk, {
    type,
    object: update
  }) {
    super(vk);
    this.payload = update;
    this.type = 'vote';
    this.subTypes = ['pull_vote'];
  }
  /**
   * Returns the identifier poll
   *
   * @return {number}
   */


  getId() {
    return this.payload.poll_id;
  }
  /**
   * Returns the identifier user
   *
   * @return {number}
   */


  getUserId() {
    return this.payload.user_id;
  }
  /**
   * Returns the identifier owner
   *
   * @return {number}
   */


  getOwnerId() {
    return this.payload.owner_id;
  }
  /**
   * Returns the identifier option
   *
   * @return {number}
   */


  getOptionId() {
    return this.payload.option_id;
  }

}

exports.default = VoteContext;
