// vk-io/lib/structures/attachments/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Attachment", {
  enumerable: true,
  get: function () {
    return _attachment.default;
  }
});
Object.defineProperty(exports, "GiftAttachment", {
  enumerable: true,
  get: function () {
    return _gift.default;
  }
});
Object.defineProperty(exports, "WallAttachment", {
  enumerable: true,
  get: function () {
    return _wall.default;
  }
});
Object.defineProperty(exports, "LinkAttachment", {
  enumerable: true,
  get: function () {
    return _link.default;
  }
});
Object.defineProperty(exports, "PhotoAttachment", {
  enumerable: true,
  get: function () {
    return _photo.default;
  }
});
Object.defineProperty(exports, "AudioAttachment", {
  enumerable: true,
  get: function () {
    return _audio.default;
  }
});
Object.defineProperty(exports, "VideoAttachment", {
  enumerable: true,
  get: function () {
    return _video.default;
  }
});
Object.defineProperty(exports, "MarketAttachment", {
  enumerable: true,
  get: function () {
    return _market.default;
  }
});
Object.defineProperty(exports, "StickerAttachment", {
  enumerable: true,
  get: function () {
    return _sticker.default;
  }
});
Object.defineProperty(exports, "DocumentAttachment", {
  enumerable: true,
  get: function () {
    return _document.default;
  }
});
Object.defineProperty(exports, "WallReplyAttachment", {
  enumerable: true,
  get: function () {
    return _wallReply.default;
  }
});
Object.defineProperty(exports, "MarketAlbumAttachment", {
  enumerable: true,
  get: function () {
    return _marketAlbum.default;
  }
});
exports.default = void 0;

var _attachment = _interopRequireDefault(require("./attachment"));

var _gift = _interopRequireDefault(require("./gift"));

var _wall = _interopRequireDefault(require("./wall"));

var _link = _interopRequireDefault(require("./link"));

var _photo = _interopRequireDefault(require("./photo"));

var _audio = _interopRequireDefault(require("./audio"));

var _video = _interopRequireDefault(require("./video"));

var _market = _interopRequireDefault(require("./market"));

var _sticker = _interopRequireDefault(require("./sticker"));

var _document = _interopRequireDefault(require("./document"));

var _wallReply = _interopRequireDefault(require("./wall-reply"));

var _marketAlbum = _interopRequireDefault(require("./market-album"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = _attachment.default;
exports.default = _default;
